### PR TITLE
Match osu-stable follow circle behaviour and size

### PR DESCRIPTION
**Pull request by thewildtree**
Brings `FollowCircle` behaviour in line with stable by introducing `InternalFollowCircle` (not sure about this name though) and making it scale instantly to `2.4f`, so the input check matches stable, while the visual part scales to `2f` on its own as it previously did.

Fixes part of #7521.

### CODE REVIEW

**Review by frenzibyte — CHANGES_REQUESTED** [review]
~~That's a little bit of a hacky way to do this, what you probably want to do is create that circular container inside the `FollowCircleContainer` **internally**.~~

**Comment by TheWildTree** [comment]
If done that way, would it let me scale the internal circle to a size bigger than the visible one (its parent)?

**Comment by bdach** [comment]
I think it would if you specified a relative `Size` bigger than 1 (1.2 in this case, I suppose). I wouldn't say the current solution is *hacky* per se, but encapsulation is better, I agree.

**Comment by TheWildTree** [comment]
Encapsulating it doesn't make sense to me, since they have to scale separately (one instantly, one over time) - if done as proposed, scaling will always be wonky for one of the circles (depending on which one we'd make the parent). Unless there is a way to decouple the scaling part, of course, but that seems counterproductive in a parent-child relationship.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 1]
```diff
@@ -38,6 +39,13 @@ public SliderBall(Slider slider, DrawableSlider drawableSlider = null)
 
             Children = new[]
             {
+                InternalFollowCircle = new CircularContainer
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: The name of this drawable is definitely misleading given what it does, how about `TrackingArea` or `FollowingArea`, with an in-line comment explaining why it's there. (and reference to osu-stable)

**Review by frenzibyte — CHANGES_REQUESTED** [review]
On second thought, this may actually be fine currently FWIW. also if possible, would be nice to add tests following the slider with the cursor beyond the follow circle to ensure no issues with this.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 2]
```diff
@@ -38,6 +39,14 @@ public SliderBall(Slider slider, DrawableSlider drawableSlider = null)
 
             Children = new[]
             {
+                // This is separate from the visible followcircle to ensure consistent internal tracking area (needed to match osu-stable)
+                TrackingArea = new CircularContainer
+                {
+                    Origin = Anchor.Centre,
+                    Anchor = Anchor.Centre,
+                    RelativeSizeAxes = Axes.Both,
+                    Alpha = 0
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: this shouldn't be required

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 3]
```diff
@@ -24,6 +24,7 @@ public class SliderBall : CircularContainer, ISliderProgress, IRequireHighFreque
 
         private readonly Slider slider;
         public readonly Drawable FollowCircle;
+        public readonly Drawable TrackingArea;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: i think this can be private

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by TheWildTree** [comment]
Added tests checking for the tracking area radius. Please let me know if that's the right way to test this - I'm a bit curious about tracking dropping on hitobjectradius * 1.2f (so theoretically on the edge of tracking area which is scaled to 2.4f), I thought it's gonna drop tracking only after passing that point (so at >1.2f), but I guess that's not the case.

**Comment by TheWildTree** [comment]
Testing this behaves differently in visual testing than it does in CLI. In visual browser, cursor `circle_radius * 1.2f` away from the center was enough to drop tracking, while in CLI it needed >1.2f to drop. 

I don't know which one was behaving correctly, so for now I decreased the precision of tests to not trigger those edge cases, as they seem to be out of scope for this PR.

**Comment by peppy** [comment]
I think that is fine for the time being. We will likely to a precision pass once we are closer to requiring this from replay playback / recording.

**Comment by peppy** [comment]
Probably important to note (though I don't think it requires a comment) that the tests are using `hit object radius * 1.2`, which is only correct because it's using default CS5

**Review by peppy — APPROVED** [review]

**Comment by smoogipoo** [comment]
What about our prior discussion where we concluded that the user should always see their input matching with something on screen? With this, not only does it not scale up but it also doesn’t match the visual representation anymore.

**Comment by peppy** [comment]
You raise a good point, completely missed that this is actually going against that.

We may want to just fix the size / tracking region to match for the time being, and revert this fix until it can be applied in a legacy-only way.

**Pull request merged**